lib/ansible/modules/packaging/os/yum.py

```diff
@@ -639,6 +639,42 @@ def local_envra(path):
                                header[rpm.RPMTAG_ARCH])
 
 
+def set_env_proxy(function):
+    def fetching(spec, module, conf_file, installroot):
+        # setting system proxy environment and saving old, if exists
+        yumb = yum_base(conf_file, installroot)
+        namepass = ""
+        schem = "http"
+        old_proxy_env = [os.getenv("http_proxy"), os.getenv("https_proxy")]
+        try:
+            if yumb.conf.proxy:
+                if 'https' in yumb.conf.proxy:
+                    schem = 'https'
+                else:
+                    schem = 'http'
+                if yumb.conf.proxy_username:
+                    namepass = namepass + yumb.conf.proxy_username
+                    if yumb.conf.proxy_password:
+                        namepass = namepass + ":" + yumb.conf.proxy_password
+                namepass = namepass + '@'
+                os.environ["{0}_proxy".format(schem)] = re.sub(
+                    r"({0}://)".format(schem), r"\1" + namepass, yumb.conf.proxy
+                )
+        except yum.Errors.YumBaseError as e:
+            module.fail_json(msg="Error setting/accessing proxy: %s" % to_native(e))
+        # getting package with new proxy environment
+        package = fetch_rpm_from_url(spec, module)
```

Comment:
If you were to do this as a decorator, you should be calling ```function``` here instead of hardcoding fetch_rpm_from_url.

Reply:
You mean using 
```
@set_env_prox(arg...)  
def fetch_rpm_from_urls(spec, module=None):
...
```
In that variant i cannot pass variables needed to YumBase 
> conf_file
> installroot

Reply:
I mean like:
``` python
-        package = fetch_rpm_from_url(spec, module)
+        package = function(spec, module)
```

But looking at this, I still think it should be a context manager rather than a pseudo-decorator.  That would look something like this:

``` python
from contextlib import contextmanager
[...]
@contextmanager
def set_env_proxy(conf_file, installroot):
    old_environ = # [old values of the proxy env vars]
    try:
        my = yum_base(conf_file, installroot)
        # [set environment to the proxies that yum knows about]
        yield
    except:
        raise
    finally:
        # reset the environment to the values saved in old_environ
[...]
if '://' in spec:
    with set_env_proxy(conf_file, installroot):
        package = fetch_rpm_from_url(spec, module=module)
[...]
```